hdate/date.py

```diff
@@ -383,6 +377,16 @@ def holiday_dates_cross_product(holiday):
             for date_instance in holiday_dates_cross_product(holiday)
             if len(holiday.date) >= 2
         ]
+        is_leap_year = self.year_size() > 380
+        # Filter any special cases defined by True/False functions
+        holidays_list = [
+            (holiday, date)
+            for (holiday, date) in holidays_list
+            if all(func(date) for func in holiday.date_functions_list)
+            # Don't return two copies of Purim.
+            if date.hdate.month
+            not in ([Months.ADAR] if is_leap_year else [Months.ADAR_I, Months.ADAR_II])
```

Comment:
Could you try and refactor this in the `htables.py`

I would add something in the following form (from htables.py):

```
    HOLIDAY(
        HolidayTypes.FAST_DAY,
        "taanit_esther",
        ([11, 13], [Months.ADAR, Months.ADAR_II]),
        "",
        [move_if_not_on_dow(13, 11, 5, 3), **leap_year()**],
        LANG("Jeûne d'Esther", "Ta'anit Esther", DESC("תענית אסתר", "תענית אסתר")),
    ),
```

And add a lambda function over there.